[bugfix] member-ordering includes getters and setters

NaridaL opened the pull request:
#### PR checklist

- [X] Addresses an existing issue: fixes #3965
- [ ] New feature, bugfix, or enhancement
- [X] Includes tests
- [ ] Documentation update

#### Overview of change:
Added categories "-accessor" for getters and setters. Added those categories to the presets. Based on https://github.com/palantir/tslint/pull/3935, merge that first.

#### Is there anything you'd like reviewers to focus on?

<!-- optional -->

#### CHANGELOG.md entry: 
[bugfix] `member-ordering` includes property accessors (getters and setters).

<!-- optional (example: "[new-rule] `arrow-return-shorthand`") -->
<!-- suggested tags: [new-rule], [new-rule-option], [new-fixer], [bugfix], [enhancement], [api], [rule-change], [no-log] -->

giladgray commented:
@NaridaL I just merged #3935 so go ahead and update this 👍

NaridaL commented:
Done. Let's see if the lint still fails...

NaridaL commented:
OK, yeah, for some reason running `npm i` locally resulted in me getting the error too.

```
ERROR: 193:5  member-ordering  Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after private instance fields.
```

The error doesn't make much sense, because tslint's tslint.json is using the old-style option "variables-before-functions", and accessors are neither. As this is going to be a breaking change, as I added the accessors to the presets, and if you didn't happen to put them in the same location you're going to get errors now. Therefore I suggest doing what the comment on line 534 suggests and remove the old-style options entirely and update tslint.json.

```
/**
 * Convert from undocumented old-style options.
 * This is designed to mimic the old behavior and should be removed eventually.
 */
```

NaridaL commented:
OK, I changed tslint's own config to use the current config type.

Can I remove the old-style option parsing entirely or do we need a version which outputs warnings first?

giladgray commented on src/language/rule/rule.ts:
```diff
@@ -165,6 +165,8 @@ export interface ReplacementJson {
     innerText: string;
 }
 export class Replacement {
+    constructor(readonly start: number, readonly length: number, readonly text: string) {}
```
why are these constructors moved? i did not expect them to come first.

NaridaL replied:
>tslint.json: use preset "fields-first" for member-ordering rule instead of previous old-style config.
>This seems to match previous config most closely, requiring only 5 changes.

I can recheck if one of the other presets matches more closely, otherwise, this seems the way to go...

NaridaL replied:
Yeah this preset is the closest.

giladgray commented on test/rules/member-ordering/alphabetize/test.ts.fix:
```diff
@@ -3,6 +3,16 @@ class X {
+    set y(x: number) {}
+
+    set z(x: number) {}
+    get z() {
```
why does it go `get set set get`? i would expect `get set get set`.

NaridaL replied:
This was discussed in the underlying issue https://github.com/palantir/tslint/issues/3965

Currently the sort works by first sorting by category, then optionally sorting alphabetically. To get 

    get x
    set x
    get y
    set y
    // other members

would require sorting first by category, then by alphabet, and a third "pass" sorting get < set. I've left this out as it would be added complexity (maybe some people want set < get) for what would only be very local errors. 

Adding this it isn't out of the question if you feel it's important.

giladgray replied:
can we interpret these as `x get, `x set`, `y get`, `y set`? that'll give the desired sorting.

NaridaL replied:
Heh, I had the exact same idea https://github.com/palantir/tslint/issues/3965#issuecomment-398552202. I dismissed it would result in things like this:
```
get x() {}
"x lol"() {}
set x() {}
```

Rethinking it however, that seems like such an unlikely scenario that it might still be a good solution.

giladgray commented on src/rules/memberOrderingRule.ts:
```diff
@@ -27,34 +27,58 @@ const OPTION_ALPHABETIZE = "alphabetize";
+
+    protectedStaticMethod,
     privateStaticMethod,
+    publicStaticMethod,
```
wrong order, should be `public,protected,private` for each type

giladgray commented on tslint.json:
```diff
@@ -49,11 +49,7 @@
-      "static-before-instance",
-      "variables-before-functions"
-    ],
+    "member-ordering": [true, { "order": "fields-first" }],
```
is the old options syntax still supported?

giladgray replied:
oh right, you asked. we cannot ship a breaking change. please preserve the old syntax, or we'll have to talk about 6.0

NaridaL commented:
@giladgray I've added new categories for accessors. I've added these to the presets. (Currently getters and setters are ignored entirely). Therefore, if a user is using one of the presets, this is going to be a breaking change one way or the other. We might as well add other breaking changes (such as removing old style options).

giladgray commented:
@NaridaL there's a big difference between a change in the lint result and an **API change** in the configuration. changing the configuration is a new API contract and _requires_ user intervention during the upgrade in a way that is not automate-able.

conversely, changing the lint result may have no effect (if it's auto-fixable, which it is in this case), or it fails their build and they can easily update the code accordingly. we do not consider lint style changes to be API breaking.

so adding new categories is not an actual API break: it's actually an enhancement. but changing the config options in a non-backward-compatible way is an API break.

NaridaL commented:
Ah, thanks for clearing up the distinction. I haven't removed the old-
style option handling code, so this pr isn't breaking by your definition. I
could add a console warning when using this option so that it could be
removed in a future release

On Tue, Jun 26, 2018, 23:33 Gilad Gray <notifications@github.com> wrote:

> @NaridaL <https://github.com/NaridaL> there's a big difference between a
> change in the lint result and an *API change* in the configuration.
> changing the configuration is a new API contract and *requires* user
> intervention during the upgrade in a way that is not automate-able.
>
> conversely, changing the lint result may have no effect (if it's
> auto-fixable, which it is in this case), or it fails their build and they
> can easily update the code accordingly. we do not consider lint style
> changes to be API breaking.
>
> so adding new categories is not an actual API break: it's actually an
> enhancement. but changing the config options in a non-backward-compatible
> way is an API break.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/palantir/tslint/pull/3984#issuecomment-400469811>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKU8SvrzB_i8io705VTp7IYvdwSydNlNks5uAqiPgaJpZM4UyIXH>
> .
>

giladgray commented:
no need to add a console warning as this would apply to _all_ rules

NaridaL commented:
@giladgray All rules?

I was thinking something like 

![image](https://user-images.githubusercontent.com/10828874/41995531-70bd96de-7a52-11e8-9a65-20807656e574.png)

giladgray commented:
that is awesome! very clear message with migration steps 👌 
however, definitely does not belong in this PR as that looks like a rather complex feature in itself and deserves to be reviewed separately.

don't think we're ready to deprecate old options usage just yet, though i imagine that'll be a flagship feature for 6.0 (that's what i meant by all rules -- migrating to object syntax across the board)

NaridaL commented:
The rationale for adding the warning here is that the old style options don't handle accessors correctly and output unhelpful warnings. This rule already handles old style options by converting them to current ones, therefore outputting the above warning doesn't require any logic.

NaridaL commented on src/rules/memberOrderingRule.ts:
```diff
@@ -469,7 +532,14 @@ function getOptionsJson(allOptions: any[]): { order: MemberCategoryJson[]; alpha
+            You should replace ${allOptions.map((o) => JSON.stringify(o)).join()}
+            with the following equivalent options and add -accessor categories as appropriate:\n` +
+            JSON.stringify(order, undefined, "  "));
+        return { order, alphabetize: false }; // presume allOptions to be string[]
```
I've pushed the warning for the previously stated reason. Let me know if you still want it in a separate PR...

NaridaL commented:
Failing tests are due to latest commit on master.

maiis commented:
@NaridaL I think you need to rebase this commit to fix the tests: 29fbace7bd67195c7e7f1f115a3170e092f4e78c

NaridaL commented:
rebased onto master

JoshuaKGoldberg commented:
Removing the `breaking change` label per #4811!

adidahiya commented:
thanks for the PR @NaridaL!
